File: helpdeskspringboot/src/main/java/com/ez/service/CategoryService.java
package com.ez.service;

import com.ez.entity.Category;
import com.ez.exception.IDNotFoundException;
import com.ez.repository.CategoryRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

import static com.ez.constant.Constant.*;

@Service
public class CategoryService{

    private Logger LOGGER = LoggerFactory.getLogger(getClass());

    @Autowired
    private CategoryRepository categoryRepository;

    // search categories by pageNumber and based on the search criteria.
    // parameters:
    //  - pageNumber: page number
    //  - pageSize: page size
    //  - searchTerm: ID, name
    //  - status: '', 'Active', 'Inactive'
    public List<Category> searchCategories(int pageNumber, int pageSize, String searchTerm, String status){

        LOGGER.info("search categories");

        return categoryRepository.searchCategories(pageNumber, pageSize, searchTerm, status);
    }

    // calculate total of categories based on the search criteria
    public long getTotalOfCategories(String searchTerm, String status) {
        LOGGER.info("get total of categories");

        return categoryRepository.getTotalOfCategories(searchTerm, status);
    }

    // create new category
    public Category createCategory(Category category){

        LOGGER.info("create new category");

        // new user
        Category newCategory = new Category(category.getName(), category.getStatus());

        // save new category into database
        categoryRepository.save(newCategory);

        return newCategory;
    }

    // find category by category id
    public Category findById(Long id) throws IDNotFoundException {

        LOGGER.info("find category by id");

        // find category by category id
        return categoryRepository.findById(id).orElseThrow(() -> new IDNotFoundException(NO_CATEGORY_FOUND_BY_ID + id));
    }

    // update existing category
    public Category updateCategory(Category category) throws IDNotFoundException {

        LOGGER.info("Update category");

        // get existing category(persistent)
        Category existingCategory = categoryRepository.findById(category.getId())
                .orElseThrow(() -> new IDNotFoundException(NO_CATEGORY_FOUND_BY_ID + category.getId()));

        // set new values to existing category
        existingCategory.setName(category.getName());
        existingCategory.setStatus(category.getStatus());

        // update existing category(persistent)
        categoryRepository.save(existingCategory);

        return existingCategory;
    }

}

File: helpdeskangular/src/app/service/category.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { Category } from '../payload/Category';

@Injectable({
  providedIn: 'root'
})
export class CategoryService {

 
  // 'http://localhost:8080'
  host = environment.apiUrl;

  // number of categories per page(default = 5)
  pageSize = environment.pageSize;

  constructor(private http: HttpClient) { }

  // get categories by page and based on the search criteria
  searchCategories(pageNumber: number, searchTerm: string, status: string): Observable<Category[]> {

    // ex: url = http://localhost:8080/category-search?pageNumber=0&pageSize=5&searchTerm=""&status=""
    // console.log(`${this.host}/category-search?pageNumber=${pageNumber}&pageSize=${this.pageSize}&searchTerm=${searchTerm}&status=${status}`)
    return this.http.get<Category[]>(
      `${this.host}/category-search?pageNumber=${pageNumber}&pageSize=${this.pageSize}&searchTerm=${searchTerm}&status=${status}`
    )
  }

  // calculate total of categories for count total pages
  getTotalOfCategories(searchTerm: string, status: string): Observable<number> {

    // ex: http://localhost:8080/total-of-categories?searchTerm=""&status=""
    return this.http.get<number>(
      `${this.host}/total-of-categories?searchTerm=${searchTerm}&status=${status}`
    );

  }

  // create new category
  public createCategory(category: Category): Observable<Category> {
    return this.http.post<Category>(`${this.host}/category-create`, category);
  }

  // edit existing category
  public editCategory(category: Category): Observable<Category> {
    return this.http.put<Category>(`${this.host}/category-edit`, category);
  }

  // find category by id
  findById(id: number): Observable<Category> {
    return this.http.get<Category>(`${this.host}/category-list/${id}`);
  }

}